pkg/k8s/watchers/endpointsynchronizer.go

```diff
@@ -140,7 +141,14 @@ func (epSync *EndpointSynchronizer) RunK8sCiliumEndpointSync(e *endpoint.Endpoin
 					}
 
 					scopedLog.Debug("Getting CEP during an initialization")
-					localCEP, err = ciliumClient.CiliumEndpoints(namespace).Get(ctx, podName, meta_v1.GetOptions{})
+					if firstTry {
+						// First we try getting CEP from the API server cache, as it's cheaper.
+						// If it fails we get it from etcd to be sure to have fresh data.
+						localCEP, err = ciliumClient.CiliumEndpoints(namespace).Get(ctx, podName, meta_v1.GetOptions{ResourceVersion: "0"})
```

Comment:
Why does doing this get from the cache?

Reply:
because of ```ResourceVersion: "0"```
https://kubernetes.io/docs/reference/using-api/api-concepts/#semantics-for-get-and-list
This is also `default` what the informer does in ListAndWatch pattern [code](https://github.com/kubernetes/kubernetes/blob/34fe27355b8b5acc4d29d053ed4361b4f72e147b/staging/src/k8s.io/client-go/tools/cache/reflector.go#L663)
It is a quite standard pattern, take a look here: https://github.com/kubernetes/kubernetes/blob/4557c694ef42da6c63f42ac4c81bea5716b471b7/pkg/controller/controller_utils.go#L1036

Normally without resourceVersion specified request goes directly to Etcd and raft consensus. With resourceVersion=0 it is answered by Kubernetes apiserver cache. This smallish improvement might make a significant improvement when cilium-agent is restarting and all agents are trying to get CEP from k8s, spamming k8s apiserver with requests.

Reply:
Makes sense to me,

I am wondering is there any specified behaviour for consistency related to separate resources of the same name (i.e. different UID).  The endpoint sync depends on CiliumEndpoint.UUID to prevent certain types of race conditions related to ambiguous "ownership" of agent/controller-instance -> CEP. There might be some weird cases if its possible for the server to return resources that no longer actually exist.

I couldn't find any specification on that from the docs.

Reply:
For example:

1. Agent restores endpoint without a CEP UID reference). Resource no longer exists, but is returned. If this is a "local" CEP (i.e. referencing a endpoint on the same Node).
    * The sync code may try to [backfill its CEP UID ref](https://github.com/cilium/cilium/blob/b91758c1496d1e489486a3143a81ab740e0b5bcc/pkg/k8s/watchers/endpointsynchronizer.go#L374) of a restored endpoint.  In this case, the CEP will have an old UID reference which will be written out to disk in the endpoint restore header.
    * Patch updates to the CEP status are [tested to ensure they have the same UID](https://github.com/cilium/cilium/blob/b91758c1496d1e489486a3143a81ab740e0b5bcc/pkg/k8s/watchers/endpointsynchronizer.go#L270).
    * In this case, the error will be `k8serrors.NotFound`(I think ??).
    * I think the EPS controller might just loop, with the wrong CEP UID, forever?

Reply:
This bit of code is particularly annoying [code](https://github.com/cilium/cilium/blob/b91758c1496d1e489486a3143a81ab740e0b5bcc/pkg/k8s/watchers/endpointsynchronizer.go#L368). It only is necessary for upgrade paths. 

Currently the only version of Cilium to guarantee writing the CEP UID to the endpoint restore header state is v1.13. So I think in v1.14 we can remove this code, this will also remove these possible annoying edge cases.

Reply:
I think if we just remove that code going forward we should be fine, that should clear up the issues I mentioned.

(I'll have to double check our upgrade path practices).

Reply:
If the patch fails, wouldn't we just zero out localCEP and retry fetching it again? https://github.com/cilium/cilium/blob/b91758c1496d1e489486a3143a81ab740e0b5bcc/pkg/k8s/watchers/endpointsynchronizer.go#L284